### ISSUE TITLE
coreutils copied to wrong locations

### ISSUE THREAD

**Issue by dothebart** [debuerreotype/docker-debian-artifacts#12]
`chmod` and `chown` plus some more utilities are copied from `/bin/` to `/usr/bin/` - which makes cmake choose a wrong path for these binaries. The packaging system clearly shows that the `/usr/bin/` version wasn't installed via a debian package:

    # dpkg -S /usr/bin/chmod
    dpkg-query: no path found matching pattern /usr/bin/chmod
    # dpkg -S /bin/chmod
    coreutils: /bin/chmod

The container should resemble native environments as closely as possible, and either not ship them at all or only in their appropriate location. 

If somebody needs them in the wrong location desperately he still can derive the `debian:stretch` container and do that copy.

However the original container shouldn't be tainted by unmanaged files that may originate anywhere.

**Comment by tianon** [comment]
This is because we enable "--merged-usr", which makes /bin a symlink to
/usr/bin (see debootstrap's man page for more details).

**Comment by dothebart** [comment]
Why isn't this the case for the ubuntu docker images then?  Whats the advantage over not doing it?

**Comment by dothebart** [comment]
and, why... 
```
docker run -it debian:jessie /bin/bash 
root@3dedb6a3e84b:/# ls -l /
total 64
drwxr-xr-x   2 root root 4096 Jan 16  2017 bin
drwxr-xr-x   2 root root 4096 Dec 28  2016 boot
drwxr-xr-x   5 root root  380 Sep  4 15:09 dev
drwxr-xr-x  41 root root 4096 Sep  4 15:09 etc
drwxr-xr-x   2 root root 4096 Dec 28  2016 home
drwxr-xr-x   9 root root 4096 Nov 27  2014 lib
drwxr-xr-x   2 root root 4096 Jan 16  2017 lib64
drwxr-xr-x   2 root root 4096 Jan 16  2017 media
drwxr-xr-x   2 root root 4096 Jan 16  2017 mnt
drwxr-xr-x   2 root root 4096 Jan 16  2017 opt
dr-xr-xr-x 249 root root    0 Sep  4 15:09 proc
drwx------   2 root root 4096 Jan 16  2017 root
drwxr-xr-x   3 root root 4096 Jan 16  2017 run
drwxr-xr-x   2 root root 4096 Jan 16  2017 sbin
drwxr-xr-x   2 root root 4096 Jan 16  2017 srv
dr-xr-xr-x  13 root root    0 Jul 20 05:56 sys
drwxrwxrwt   2 root root 4096 Jan 16  2017 tmp
drwxr-xr-x  10 root root 4096 Jan 16  2017 usr
drwxr-xr-x  11 root root 4096 Jan 16  2017 var
root@3dedb6a3e84b:/# ls -l /etc/alternatives/
total 4
-rw-r--r-- 1 root root 100 May  2  2016 README
lrwxrwxrwx 1 root root  13 Jan 16  2017 awk -> /usr/bin/mawk
lrwxrwxrwx 1 root root  29 Jan 16  2017 awk.1.gz -> /usr/share/man/man1/mawk.1.gz
lrwxrwxrwx 1 root root  38 Jan 16  2017 builtins.7.gz -> /usr/share/man/man7/bash-builtins.7.gz
lrwxrwxrwx 1 root root  13 Jan 16  2017 nawk -> /usr/bin/mawk
lrwxrwxrwx 1 root root  29 Jan 16  2017 nawk.1.gz -> /usr/share/man/man1/mawk.1.gz
lrwxrwxrwx 1 root root   9 Jan 16  2017 pager -> /bin/more
lrwxrwxrwx 1 root root  29 Jan 16  2017 pager.1.gz -> /usr/share/man/man1/more.1.gz
lrwxrwxrwx 1 root root  17 Jan 16  2017 rmt -> /usr/sbin/rmt-tar
lrwxrwxrwx 1 root root  32 Jan 16  2017 rmt.8.gz -> /usr/share/man/man8/rmt-tar.8.gz
lrwxrwxrwx 1 root root  17 Jan 16  2017 w -> /usr/bin/w.procps
lrwxrwxrwx 1 root root  33 Jan 16  2017 w.1.gz -> /usr/share/man/man1/w.procps.1.gz
```
?

**Comment by ObiWahn** [comment]
Please stick the the behavior of the distribution. And enable it once it becomes the debian default. This is meant to avoid build problems when building and to create more compatibility between distributions. But when your build (created in the docker image) is not compatible to default installations, then you have gained nothing. We face the problem that ch(own|mod) is found in /usr/bin at build time but not in default installations. 

https://github.com/arangodb/arangodb/blob/devel/CMakeLists.txt#L288

**Comment by tianon** [comment]
> Why isn't this the case for the ubuntu docker images then?

Because Canonical has not chosen to enable it when they invoke `debootstrap` to create their Docker rootfs tarballs.

> Whats the advantage over not doing it?

Debian (the `debootstrap` maintainers / `debian-boot` team in particular) is actively working to switch the default to `--merged-usr`, and it has been enabled in the Docker images for quite some time now (specifically, since https://github.com/docker-library/official-images/pull/2179, which was Sep 23, 2016).

One issue is that we've also applied `--merged-usr` on Jessie, which didn't yet fully support it (and that occurred when we adjusted these Debian tarballs to be created reproducibly via `debuerreotype`), where with the previous method of building the images, we did not.  However, [the upstream build script already excludes `jessie`, which is why the symlinks are not seen there](https://anonscm.debian.org/cgit/d-i/debootstrap.git/tree/scripts/sid?id=70f02cb77d3a5238d3f18c88ec4d4d15eadb9dc2#n45) (and so our erroneous inclusion of the flag while building `jessie` is harmless).

For Stretch and above, this change is almost a year old now, and was a direct request from the `debootstrap` maintainers explicitly for us to enable in our Docker images (see https://github.com/docker-library/official-images/pull/2179#issuecomment-249257219), and as such will not be reverted.

If I can get a simple `Dockerfile` which reproduces any issues due to this, I'd be happy to take a look and see what I can do to help fix them.

**Comment by ObiWahn** [comment]
merged-usr is the furture but should not be enabled by others before the distribution enables it!

Most people really not care about debootstrap what counts is the default installation from official images that you can download or that are available in hoster's installation menus. A mismatch between those and build containers is whack. Debootstrap is for sure relevant for servers and I admit that I use it quite often with systemd-nspawn. But why do they ask you to pass the flag when it is a safe default? Shouldn't it be a default debootstrap option requiring no action on your side.

Having **/usr/bin/chmod** in a build container plus having it probably first in PATH while it is **not available in a default installation** is just broken. It confuses tools like cmake in a way that macros like `find_program` will not work anymore. So you are forcing people that need to maintain the build for several operating systems to include extra fixes.

I really fail to see any advantage in that! Is building / testing / deploying software what docker is mostly used for? Or do people use docker to create some really cool images that they can use to serve "bin" to the network from a single directory?

**Comment by tianon** [comment]
I talked this over with several other DDs (including the maintainer who'd asked for us to enable it in the first place), and the consensus is to disable it for `stretch` (since it's now released), and to leave it enabled for `buster` and `sid` for now, re-evaluating that closer to `buster`'s release time (based on https://bugs.debian.org/858331 / https://bugs.debian.org/cgi-bin/848622).

I'll work on a PR to fix the code today, although I can't make any promises about how quickly we'll get an updated image out (especially since once it's fixed, we need to rebuild ~12 suites times 7 architectures, and it then triggers a ton of automated builds and official images rebuilds, but that being said, we are due for a Debian image bump regardless).

TL;DR, we're going to fix this, and said fix should be available sometime this week.

**Comment by tianon** [comment]
See https://github.com/debuerreotype/debuerreotype/pull/20 for the `debuerreotype` PR which will trickle down to being the fix for this.

**Comment by dothebart** [comment]
Thanks a lot. If a native Buster install will behave similar I'm totally fine with it!

**Comment by tianon** [comment]
Updated image is pushed now with https://github.com/docker-library/official-images/pull/3425. :+1:
